File: skills/codex-app-changelog/scripts/fetch_latest_app_changelog.py
#!/usr/bin/env python3
"""Fetch the latest Codex app changelog entry from OpenAI Developers."""
from __future__ import annotations

import argparse
import datetime as dt
import json
import re
import sys
from dataclasses import dataclass
from html.parser import HTMLParser
from pathlib import Path
from typing import Iterable
from urllib.error import URLError
from urllib.request import Request, urlopen


SOURCE_URL = "https://developers.openai.com/codex/changelog?type=codex-app"
BASE_SOURCE_URL = "https://developers.openai.com/codex/changelog"
TOPIC = "codex-app"
SKILL_DIR = Path(__file__).resolve().parent.parent
DEFAULT_OUT_DIR = SKILL_DIR / "changelogs"
VERSION_RE = re.compile(r"\b(\d{2}\.\d{3}(?:\.\d+)?)\b")


class FetchError(RuntimeError):
    pass


def comparable_markdown(text: str) -> str:
    return "\n".join(
        line for line in text.replace("\r\n", "\n").replace("\r", "\n").splitlines()
        if not line.startswith("fetched:")
    ).strip()


@dataclass
class Entry:
    entry_id: str
    topics: list[str]
    published: str
    title: str
    body: str


def collapse_space(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()


def source_for_entry(entry_id: str) -> str:
    return f"{BASE_SOURCE_URL}#{entry_id}"


def version_from_title(title: str, entry_id: str) -> str:
    versions = VERSION_RE.findall(title)
    if versions:
        return "-".join(versions)
    if entry_id:
        return entry_id.removeprefix("codex-")
    return dt.date.today().isoformat()


def add_text(buffer: list[str], data: str) -> None:
    text = collapse_space(data)
    if not text:
        return
    previous = buffer[-1] if buffer else ""
    if (
        buffer
        and previous not in {"`", "**", "*"}
        and not previous.endswith((" ", "\n"))
        and not text.startswith((".", ",", ":", ";", ")", "]"))
    ):
        buffer.append(" ")
    buffer.append(text)


def add_inline_marker(buffer: list[str], marker: str, *, closing: bool) -> None:
    if closing and buffer:
        buffer[-1] = f"{buffer[-1]}{marker}"
        return
    if buffer and not buffer[-1].endswith((" ", "\n", "(", "[")):
        buffer.append(" ")
    buffer.append(marker)


class CodexChangelogParser(HTMLParser):
    def __init__(self, topic: str) -> None:
        super().__init__(convert_charrefs=True)
        self.topic = topic
        self.entries: list[Entry] = []

        self.in_entry = False
        self.entry_depth = 0
        self.entry_matches = False
        self.entry_id = ""
        self.entry_topics: list[str] = []

        self.in_article = False
        self.article_depth = 0
        self.time_depth = 0
        self.title_depth = 0
        self.heading_depth = 0
        self.para_depth = 0
        self.list_depth = 0
        self.item_depth = 0

        self.published: list[str] = []
        self.title: list[str] = []
        self.article_lines: list[str] = []
        self.heading_text: list[str] = []
        self.para_text: list[str] = []
        self.item_text: list[str] = []

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        attrs_dict = {key: value or "" for key, value in attrs}

        if tag == "li" and "data-codex-topics" in attrs_dict:
            self.in_entry = True
            self.entry_depth = 1
            self.entry_topics = attrs_dict.get("data-codex-topics", "").split()
            self.entry_matches = self.topic in self.entry_topics
            self.entry_id = attrs_dict.get("id", "")
            self.published = []
            self.title = []
            self.article_lines = []
            return

        if not self.in_entry:
            return

        self.entry_depth += 1

        if not self.entry_matches:
            return

        if self.in_article:
            self.article_depth += 1
            self.handle_article_start(tag, attrs_dict)
            return

        if tag == "article":
            self.in_article = True
            self.article_depth = 1
            return
        if tag == "time":
            self.time_depth += 1
            return
        if tag == "h3":
            self.title_depth += 1

    def handle_endtag(self, tag: str) -> None:
        if not self.in_entry:
            return

        if self.entry_matches:
            if self.in_article:
                self.handle_article_end(tag)
                self.article_depth -= 1
                if self.article_depth == 0:
                    self.in_article = False
            elif tag == "time" and self.time_depth:
                self.time_depth -= 1
            elif tag == "h3" and self.title_depth:
                self.title_depth -= 1

        self.entry_depth -= 1
        if self.entry_depth == 0:
            if self.entry_matches:
                body = "\n".join(self.trim_blank_lines(self.article_lines)).strip()
                self.entries.append(
                    Entry(
                        entry_id=self.entry_id,
                        topics=self.entry_topics,
                        published=collapse_space("".join(self.published)),
                        title=collapse_space("".join(self.title)),
                        body=body,
                    )
                )
            self.in_entry = False
            self.entry_matches = False

    def handle_data(self, data: str) -> None:
        if not self.in_entry or not self.entry_matches:
            return

        if self.time_depth:
            add_text(self.published, data)
            return
        if self.title_depth and not self.in_article:
            add_text(self.title, data)
            return
        if self.in_article:
            self.add_article_text(data)

    def handle_article_start(self, tag: str, attrs: dict[str, str]) -> None:
        if tag in {"h2", "h3", "h4", "h5", "h6"}:
            self.heading_depth += 1
            self.heading_text = []
        elif tag == "p":
            self.para_depth += 1
            self.para_text = []
        elif tag in {"ul", "ol"}:
            self.list_depth += 1
        elif tag == "li":
            self.item_depth += 1
            self.item_text = []
        elif tag == "br":
            self.add_article_text("\n")
        elif tag == "code":
            self.add_article_marker("`", closing=False)
        elif tag in {"strong", "b"}:
            self.add_article_marker("**", closing=False)
        elif tag in {"em", "i"}:
            self.add_article_marker("*", closing=False)

    def handle_article_end(self, tag: str) -> None:
        if tag in {"strong", "b"}:
            self.add_article_marker("**", closing=True)
        elif tag in {"em", "i"}:
            self.add_article_marker("*", closing=True)
        elif tag == "code":
            self.add_article_marker("`", closing=True)

        if tag in {"h2", "h3", "h4", "h5", "h6"} and self.heading_depth:
            heading = collapse_space("".join(self.heading_text))
            if heading:
                self.append_block(f"## {heading}")
            self.heading_depth -= 1
            self.heading_text = []
        elif tag == "p" and self.para_depth:
            paragraph = collapse_space("".join(self.para_text))
            if paragraph:
                self.append_block(paragraph)
            self.para_depth -= 1
            self.para_text = []
        elif tag == "li" and self.item_depth:
            item = collapse_space("".join(self.item_text))
            if item:
                self.article_lines.append(f"- {item}")
            self.item_depth -= 1
            self.item_text = []
        elif tag in {"ul", "ol"} and self.list_depth:
            self.list_depth -= 1
            self.append_blank()

    def add_article_text(self, data: str) -> None:
        if self.heading_depth:
            add_text(self.heading_text, data)
        elif self.item_depth:
            add_text(self.item_text, data)
        elif self.para_depth:
            add_text(self.para_text, data)

    def add_article_marker(self, marker: str, *, closing: bool) -> None:
        if self.heading_depth:
            add_inline_marker(self.heading_text, marker, closing=closing)
        elif self.item_depth:
            add_inline_marker(self.item_text, marker, closing=closing)
        elif self.para_depth:
            add_inline_marker(self.para_text, marker, closing=closing)

    def append_block(self, text: str) -> None:
        self.append_blank()
        self.article_lines.append(text)
        self.append_blank()

    def append_blank(self) -> None:
        if self.article_lines and self.article_lines[-1] != "":
            self.article_lines.append("")

    @staticmethod
    def trim_blank_lines(lines: Iterable[str]) -> list[str]:
        trimmed = list(lines)
        while trimmed and not trimmed[0]:
            trimmed.pop(0)
        while trimmed and not trimmed[-1]:
            trimmed.pop()
        return trimmed


def fetch_html(url: str) -> str:
    request = Request(url, headers={"User-Agent": "Mozilla/5.0"})
    try:
        with urlopen(request, timeout=60) as response:
            return response.read().decode("utf-8")
    except URLError as exc:
        raise FetchError(f"Failed to fetch {url}: {exc}") from exc


def fetch_entries(url: str, topic: str) -> list[Entry]:
    parser = CodexChangelogParser(topic=topic)
    parser.feed(fetch_html(url))
    return parser.entries


def select_entry(entries: list[Entry], entry_id: str | None) -> Entry:
    if not entries:
        raise FetchError(f"No Codex app changelog entries found for topic `{TOPIC}`.")
    if not entry_id:
        return entries[0]
    for entry in entries:
        if entry.entry_id == entry_id:
            return entry
    raise FetchError(f"No Codex app changelog entry found with id `{entry_id}`.")


def markdown_for_entry(entry: Entry) -> str:
    version = version_from_title(entry.title, entry.entry_id)
    fetched = dt.date.today().isoformat()
    source = source_for_entry(entry.entry_id)
    body = entry.body.strip()

    frontmatter = [
        "---",
        f"version: {json.dumps(version)}",
        f"entry_id: {json.dumps(entry.entry_id)}",
        f"title: {json.dumps(entry.title, ensure_ascii=False)}",
        f"source: {source}",
        f"published: {json.dumps(entry.published)}",
        f"fetched: {fetched}",
        "language: en",
        f"topic: {TOPIC}",
        "---",
        "",
        f"# Codex App {version}",
        "",
        f"## {entry.title}",
        "",
        body,
        "",
    ]
    return "\n".join(frontmatter)


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--save", action="store_true", help="Save markdown to changelogs/<version>.md")
    ap.add_argument("--output-dir", default=str(DEFAULT_OUT_DIR), help="Directory for --save output")
    ap.add_argument("--json", action="store_true", help="Print selected entry metadata as JSON")
    ap.add_argument("--entry-id", default=None, help="Fetch a specific changelog entry id")
    ap.add_argument("--url", default=SOURCE_URL, help="Codex changelog URL")
    args = ap.parse_args()
    out_dir = Path(args.output_dir).expanduser().resolve()

    try:
        entry = select_entry(fetch_entries(args.url, TOPIC), args.entry_id)
    except FetchError as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        return 2

    version = version_from_title(entry.title, entry.entry_id)
    markdown = markdown_for_entry(entry)

    if args.json:
        payload = {
            "version": version,
            "entry_id": entry.entry_id,
            "title": entry.title,
            "source": source_for_entry(entry.entry_id),
            "published": entry.published,
            "topics": entry.topics,
        }
        out_path = out_dir / f"{version}.md"
        if out_path.exists():
            existing = out_path.read_text(encoding="utf-8")
            payload["existing_local_path"] = str(out_path)
            payload["archive_status"] = (
                "unchanged"
                if comparable_markdown(existing) == comparable_markdown(markdown)
                else "different"
            )
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    else:
        print(markdown)

    if args.save:
        out_dir.mkdir(parents=True, exist_ok=True)
        out_path = out_dir / f"{version}.md"
        if out_path.exists():
            existing = out_path.read_text(encoding="utf-8")
            if comparable_markdown(existing) == comparable_markdown(markdown):
                print(f"Skipped unchanged: {out_path}", file=sys.stderr)
                return 0
        out_path.write_text(markdown, encoding="utf-8")
        print(f"Saved: {out_path}", file=sys.stderr)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: skills/codex-app-changelog/scripts/fetch_latest_app_changelog.ts
#!/usr/bin/env -S npx --yes tsx
/**
 * Fetch the latest Codex App changelog entry from OpenAI Developers.
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { execFileSync, spawnSync } from "node:child_process";

const SOURCE_URL = "https://developers.openai.com/codex/changelog?type=codex-app";
const BASE_SOURCE_URL = "https://developers.openai.com/codex/changelog";
const TOPIC = "codex-app";
const SKILL_NAME = "codex-app-changelog";
const VERSION_RE = /\b(\d{2}\.\d{3}(?:\.\d+)?)\b/g;

interface Entry {
  entryId: string;
  topics: string[];
  published: string;
  title: string;
  body: string;
}

interface Args {
  save: boolean;
  outputDir: string;
  json: boolean;
  entryId: string | null;
  url: string;
}

class FetchError extends Error {}

function projectRoot(): string {
  try {
    const root = execFileSync("git", ["rev-parse", "--show-toplevel"], {
      cwd: process.cwd(),
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
      timeout: 5000,
    }).trim();
    return root ? path.resolve(root) : process.cwd();
  } catch {
    return process.cwd();
  }
}

function defaultRuntimeDir(): string {
  return path.join(projectRoot(), ".context", SKILL_NAME);
}

function defaultOutputDir(): string {
  return path.join(defaultRuntimeDir(), "changelogs");
}

function ensureRuntimeGitignore(): void {
  const runtimeDir = defaultRuntimeDir();
  fs.mkdirSync(runtimeDir, { recursive: true });
  const gitignore = path.join(runtimeDir, ".gitignore");
  if (fs.existsSync(gitignore)) return;
  fs.writeFileSync(
    gitignore,
    [
      "assets/rendered/",
      "assets/covers/",
      "*.png",
      "*.jpg",
      "*.jpeg",
      "*.webp",
      "*.gif",
      "*.mp4",
      "*.mov",
      ".DS_Store",
      "__pycache__/",
      "",
    ].join("\n"),
    "utf8",
  );
}

function isInside(child: string, parent: string): boolean {
  const relative = path.relative(parent, child);
  return relative === "" || (!!relative && !relative.startsWith("..") && !path.isAbsolute(relative));
}

function parseArgs(argv: string[]): Args {
  const args: Args = {
    save: false,
    outputDir: defaultOutputDir(),
    json: false,
    entryId: null,
    url: SOURCE_URL,
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--save") args.save = true;
    else if (arg === "--json") args.json = true;
    else if (arg === "--entry-id") args.entryId = argv[++i] ?? null;
    else if (arg === "--url") args.url = argv[++i] ?? SOURCE_URL;
    else if (arg === "--output-dir") args.outputDir = expandHome(argv[++i] ?? "");
    else if (arg === "--help" || arg === "-h") {
      printHelp();
      process.exit(0);
    } else {
      throw new FetchError(`Unknown argument: ${arg}`);
    }
  }
  return args;
}

function printHelp(): void {
  process.stdout.write(`Usage: fetch_latest_app_changelog.ts [--json] [--save] [--output-dir DIR] [--entry-id ID] [--url URL]\n`);
}

function expandHome(value: string): string {
  if (value === "~") return process.env.HOME ?? value;
  if (value.startsWith("~/")) return path.join(process.env.HOME ?? "", value.slice(2));
  return value;
}

function comparableMarkdown(text: string): string {
  return text
    .replace(/\r\n/g, "\n")
    .replace(/\r/g, "\n")
    .split("\n")
    .filter((line) => !line.startsWith("fetched:"))
    .join("\n")
    .trim();
}

function collapseSpace(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

function decodeHtml(text: string): string {
  return text.replace(/&(#x?[0-9a-fA-F]+|[a-zA-Z]+);/g, (match, entity) => {
    if (entity.startsWith("#x")) return String.fromCodePoint(parseInt(entity.slice(2), 16));
    if (entity.startsWith("#")) return String.fromCodePoint(parseInt(entity.slice(1), 10));
    const named: Record<string, string> = {
      amp: "&",
      lt: "<",
      gt: ">",
      quot: '"',
      apos: "'",
      nbsp: " ",
      mdash: "—",
      ndash: "–",
      hellip: "…",
    };
    return named[entity] ?? match;
  });
}

function addText(buffer: string[], data: string): void {
  const text = collapseSpace(decodeHtml(data));
  if (!text) return;
  const previous = buffer.at(-1) ?? "";
  if (
    buffer.length > 0 &&
    !["`", "**", "*"].includes(previous) &&
    !/[ \n]$/.test(previous) &&
    !/^[.,:;)\]]/.test(text)
  ) {
    buffer.push(" ");
  }
  buffer.push(text);
}

function addInlineMarker(buffer: string[], marker: string, closing: boolean): void {
  if (closing && buffer.length > 0) {
    buffer[buffer.length - 1] = `${buffer[buffer.length - 1]}${marker}`;
    return;
  }
  if (buffer.length > 0 && !/[ \n([]$/.test(buffer[buffer.length - 1])) buffer.push(" ");
  buffer.push(marker);
}

function parseAttrs(rawTag: string): Record<string, string> {
  const body = rawTag
    .replace(/^<\s*\/?\s*[^\s/>]+/, "")
    .replace(/\/?\s*>$/, "");
  const attrs: Record<string, string> = {};
  const re = /([\w:-]+)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s"'=<>`]+)))?/g;
  let match: RegExpExecArray | null;
  while ((match = re.exec(body))) {
    attrs[match[1]] = decodeHtml(match[2] ?? match[3] ?? match[4] ?? "");
  }
  return attrs;
}

function trimBlankLines(lines: string[]): string[] {
  const result = [...lines];
  while (result.length && !result[0]) result.shift();
  while (result.length && !result[result.length - 1]) result.pop();
  return result;
}

const VOID_TAGS = new Set(["area", "base", "br", "col", "embed", "hr", "img", "input", "link", "meta", "param", "source", "track", "wbr"]);

class CodexChangelogParser {
  entries: Entry[] = [];
  private inEntry = false;
  private entryDepth = 0;
  private entryMatches = false;
  private entryId = "";
  private entryTopics: string[] = [];
  private inArticle = false;
  private articleDepth = 0;
  private timeDepth = 0;
  private titleDepth = 0;
  private headingDepth = 0;
  private paraDepth = 0;
  private listDepth = 0;
  private itemDepth = 0;
  private published: string[] = [];
  private title: string[] = [];
  private articleLines: string[] = [];
  private headingText: string[] = [];
  private paraText: string[] = [];
  private itemText: string[] = [];

  constructor(private topic: string) {}

  parse(html: string): Entry[] {
    const tokens = html.match(/<[^>]+>|[^<]+/g) ?? [];
    for (const token of tokens) {
      if (token.startsWith("</")) this.handleEnd(token);
      else if (token.startsWith("<")) this.handleStart(token);
      else this.handleData(token);
    }
    return this.entries;
  }

  private handleStart(rawTag: string): void {
    const tag = this.tagName(rawTag);
    if (!tag) return;
    const attrs = parseAttrs(rawTag);
    const isVoid = VOID_TAGS.has(tag) || rawTag.endsWith("/>");

    if (tag === "li" && "data-codex-topics" in attrs) {
      this.inEntry = true;
      this.entryDepth = 1;
      this.entryTopics = attrs["data-codex-topics"].split(/\s+/).filter(Boolean);
      this.entryMatches = this.entryTopics.includes(this.topic);
      this.entryId = attrs.id ?? "";
      this.published = [];
      this.title = [];
      this.articleLines = [];
      return;
    }

    if (!this.inEntry) return;
    if (!isVoid) this.entryDepth += 1;
    if (!this.entryMatches) return;

    if (this.inArticle) {
      if (!isVoid) this.articleDepth += 1;
      this.handleArticleStart(tag);
      return;
    }

    if (tag === "article") {
      this.inArticle = true;
      this.articleDepth = 1;
    } else if (tag === "time") {
      this.timeDepth += 1;
    } else if (tag === "h3") {
      this.titleDepth += 1;
    }
  }

  private handleEnd(rawTag: string): void {
    const tag = this.tagName(rawTag);
    if (!tag || !this.inEntry) return;

    if (this.entryMatches) {
      if (this.inArticle) {
        this.handleArticleEnd(tag);
        this.articleDepth -= 1;
        if (this.articleDepth === 0) this.inArticle = false;
      } else if (tag === "time" && this.timeDepth) {
        this.timeDepth -= 1;
      } else if (tag === "h3" && this.titleDepth) {
        this.titleDepth -= 1;
      }
    }

    this.entryDepth -= 1;
    if (this.entryDepth === 0) {
      if (this.entryMatches) {
        this.entries.push({
          entryId: this.entryId,
          topics: this.entryTopics,
          published: collapseSpace(this.published.join("")),
          title: collapseSpace(this.title.join("")),
          body: trimBlankLines(this.articleLines).join("\n").trim(),
        });
      }
      this.inEntry = false;
      this.entryMatches = false;
    }
  }

  private handleData(data: string): void {
    if (!this.inEntry || !this.entryMatches) return;
    if (this.timeDepth) addText(this.published, data);
    else if (this.titleDepth && !this.inArticle) addText(this.title, data);
    else if (this.inArticle) this.addArticleText(data);
  }

  private handleArticleStart(tag: string): void {
    if (/^h[2-6]$/.test(tag)) {
      this.headingDepth += 1;
      this.headingText = [];
    } else if (tag === "p") {
      this.paraDepth += 1;
      this.paraText = [];
    } else if (tag === "ul" || tag === "ol") {
      this.listDepth += 1;
    } else if (tag === "li") {
      this.itemDepth += 1;
      this.itemText = [];
    } else if (tag === "br") {
      this.addArticleText("\n");
    } else if (tag === "code") {
      this.addArticleMarker("`", false);
    } else if (tag === "strong" || tag === "b") {
      this.addArticleMarker("**", false);
    } else if (tag === "em" || tag === "i") {
      this.addArticleMarker("*", false);
    }
  }

  private handleArticleEnd(tag: string): void {
    if (tag === "strong" || tag === "b") this.addArticleMarker("**", true);
    else if (tag === "em" || tag === "i") this.addArticleMarker("*", true);
    else if (tag === "code") this.addArticleMarker("`", true);

    if (/^h[2-6]$/.test(tag) && this.headingDepth) {
      const heading = collapseSpace(this.headingText.join(""));
      if (heading) this.appendBlock(`## ${heading}`);
      this.headingDepth -= 1;
      this.headingText = [];
    } else if (tag === "p" && this.paraDepth) {
      const paragraph = collapseSpace(this.paraText.join(""));
      if (paragraph) this.appendBlock(paragraph);
      this.paraDepth -= 1;
      this.paraText = [];
    } else if (tag === "li" && this.itemDepth) {
      const item = collapseSpace(this.itemText.join(""));
      if (item) this.articleLines.push(`- ${item}`);
      this.itemDepth -= 1;
      this.itemText = [];
    } else if ((tag === "ul" || tag === "ol") && this.listDepth) {
      this.listDepth -= 1;
      this.appendBlank();
    }
  }

  private addArticleText(data: string): void {
    if (this.headingDepth) addText(this.headingText, data);
    else if (this.itemDepth) addText(this.itemText, data);
    else if (this.paraDepth) addText(this.paraText, data);
  }

  private addArticleMarker(marker: string, closing: boolean): void {
    if (this.headingDepth) addInlineMarker(this.headingText, marker, closing);
    else if (this.itemDepth) addInlineMarker(this.itemText, marker, closing);
    else if (this.paraDepth) addInlineMarker(this.paraText, marker, closing);
  }

  private appendBlock(text: string): void {
    this.appendBlank();
    this.articleLines.push(text);
    this.appendBlank();
  }

  private appendBlank(): void {
    if (this.articleLines.length && this.articleLines[this.articleLines.length - 1] !== "") {
      this.articleLines.push("");
    }
  }

  private tagName(rawTag: string): string {
    return rawTag.match(/^<\s*\/?\s*([a-zA-Z0-9]+)/)?.[1].toLowerCase() ?? "";
  }
}

async function fetchHtml(url: string): Promise<string> {
  const curl = spawnSync("curl", ["-fsSL", "-A", "Mozilla/5.0", url], {
    encoding: "utf8",
    timeout: 60_000,
    maxBuffer: 20 * 1024 * 1024,
  });
  if (!curl.error && curl.status === 0 && curl.stdout) return curl.stdout;

  const response = await fetch(url, {
    headers: {
      "user-agent": "Mozilla/5.0",
      accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
      "accept-language": "en-US,en;q=0.9",
    },
  });
  if (!response.ok) {
    const curlError = curl.error ? curl.error.message : (curl.stderr || curl.stdout || "").trim();
    throw new FetchError(`Failed to fetch ${url}: HTTP ${response.status}${curlError ? `; curl said: ${curlError}` : ""}`);
  }
  return response.text();
}

async function fetchEntries(url: string, topic: string): Promise<Entry[]> {
  return new CodexChangelogParser(topic).parse(await fetchHtml(url));
}

function selectEntry(entries: Entry[], entryId: string | null): Entry {
  if (!entries.length) throw new FetchError(`No Codex app changelog entries found for topic \`${TOPIC}\`.`);
  if (!entryId) return entries[0];
  const entry = entries.find((item) => item.entryId === entryId);
  if (!entry) throw new FetchError(`No Codex app changelog entry found with id \`${entryId}\`.`);
  return entry;
}

function sourceForEntry(entryId: string): string {
  return `${BASE_SOURCE_URL}#${entryId}`;
}

function today(): string {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")}`;
}

function versionFromTitle(title: string, entryId: string): string {
  const versions = [...title.matchAll(VERSION_RE)].map((m) => m[1]);
  if (versions.length) return versions.join("-");
  if (entryId) return entryId.replace(/^codex-/, "");
  return today();
}

function markdownForEntry(entry: Entry): string {
  const version = versionFromTitle(entry.title, entry.entryId);
  const body = entry.body.trim();
  return [
    "---",
    `version: ${JSON.stringify(version)}`,
    `entry_id: ${JSON.stringify(entry.entryId)}`,
    `title: ${JSON.stringify(entry.title)}`,
    `source: ${sourceForEntry(entry.entryId)}`,
    `published: ${JSON.stringify(entry.published)}`,
    `fetched: ${today()}`,
    "language: en",
    `topic: ${TOPIC}`,
    "---",
    "",
    `# Codex App ${version}`,
    "",
    `## ${entry.title}`,
    "",
    body,
    "",
  ].join("\n");
}

async function main(): Promise<number> {
  const args = parseArgs(process.argv.slice(2));
  const outDir = path.resolve(expandHome(args.outputDir));
  const entry = selectEntry(await fetchEntries(args.url, TOPIC), args.entryId);
  const version = versionFromTitle(entry.title, entry.entryId);
  const markdown = markdownForEntry(entry);
  const outPath = path.join(outDir, `${version}.md`);

  if (args.json) {
    const payload: Record<string, unknown> = {
      version,
      entry_id: entry.entryId,
      title: entry.title,
      source: sourceForEntry(entry.entryId),
      published: entry.published,
      topics: entry.topics,
    };
    if (fs.existsSync(outPath)) {
      const existing = fs.readFileSync(outPath, "utf8");
      payload.existing_local_path = outPath;
      payload.archive_status = comparableMarkdown(existing) === comparableMarkdown(markdown) ? "unchanged" : "different";
    }
    process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`);
  } else {
    process.stdout.write(markdown);
  }

  if (args.save) {
    if (isInside(outDir, defaultRuntimeDir())) ensureRuntimeGitignore();
    fs.mkdirSync(outDir, { recursive: true });
    if (fs.existsSync(outPath)) {
      const existing = fs.readFileSync(outPath, "utf8");
      if (comparableMarkdown(existing) === comparableMarkdown(markdown)) {
        process.stderr.write(`Skipped unchanged: ${outPath}\n`);
        return 0;
      }
    }
    fs.writeFileSync(outPath, markdown, "utf8");
    process.stderr.write(`Saved: ${outPath}\n`);
  }

  return 0;
}

main().then(
  (code) => process.exit(code),
  (error) => {
    process.stderr.write(`ERROR: ${error instanceof Error ? error.message : String(error)}\n`);
    process.exit(2);
  },
);
